### PR TITLE
[WIP] Add new restrict_vocab functionality, most_similar_among

### PR DESCRIPTION
**Pull request by shubhvachher**
Fix #481 .

I have looked at the suggestions there and implemented this fix in a new function most_similar_among. @gojomo. I have tested the results and they work well.

The topn variable there has a new functionality that became necessary due to the restricted vocabulary.

Also added being able to get vocabulary as a list from a trained word2vec model.

After the first review, will work on making other _among functions.

### CODE REVIEW

File: gensim/models/keyedvectors.py [hunk 1]
```diff
@@ -336,6 +346,95 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+        elif isinstance(restrict_vocab, list) or isinstance(restrict_vocab, set):
+            restrict_vocab = set(restrict_vocab)
+            vocabulary_words = set(self.get_words_from_vocab())
+
+            words_to_use = vocabulary_words.intersection(restrict_vocab)
+
+            if(not suppress_warning):
+                toWarn = "The following words are not in trained vocabulary : "
+                toWarn += str(restrict_vocab.difference(vocabulary_words))
+                warnings.warn(toWarn, UserWarning)
```
Comment (gojomo) [inline gensim/models/keyedvectors.py]: Seems this will log an attention-grabbing WARNING even if all words are present. Preferable to only log when problems occur.

File: gensim/models/keyedvectors.py [hunk 2]
```diff
@@ -336,6 +346,95 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+            if(not words_to_use):
+                raise ValueError("None of the words in restrict_vocab exist in current vocabulary")
+
+            restrict_vocab_indices = [self.vocab[word].index for word in words_to_use]
+            limited = self.syn0norm[restrict_vocab_indices] #syn0norm is an ndarray
```
Comment (gojomo) [inline gensim/models/keyedvectors.py]: I think this creates a new array (rather than just a view), which might be a significant memory cost, for large models and when the words-of-interest are a large subset of all words. Maybe this is unavoidable, but might there be any way to do the calculation against the indexes in the original  `syn0norm`?

File: gensim/models/word2vec.py [hunk 3]
```diff
@@ -1180,9 +1180,16 @@ def intersect_word2vec_format(self, fname, lockf=0.0, binary=False, encoding='ut
                         self.wv.syn0[self.wv.vocab[word].index] = weights
         logger.info("merged %d vectors into %s matrix from %s" % (overlap_count, self.wv.syn0.shape, fname))
 
+    def get_words_from_vocab(self):
```
Comment (gojomo) [inline gensim/models/word2vec.py]: For new KeyedVectors functionality, I don't think we need to add forwarding methods to Word2Vec – those exist to ease the way for older code; any new code can access the KeyedVectors methods directly. (This goes for this method and `most_similar_among()`.)

File: gensim/models/keyedvectors.py [hunk 4]
```diff
@@ -272,6 +273,15 @@ def word_vec(self, word, use_norm=False):
         else:
             raise KeyError("word '%s' not in vocabulary" % word)
 
+    def get_words_from_vocab(self):
```
Comment (shubhvachher) [inline gensim/models/keyedvectors.py]: Oh, yes! Will change! I'm still keeping the get_words_from_vocab() because when I started out with gensim, it was confusing for me initially what the w2v.wv.vocab keys and items were for!
```python
    def get_words_from_vocab(self):
        """
        returns the words in the current vocabulary as a list.
        """

        if(not self.index2word):
            raise ValueError("Vocabulary needs to be built before calling this function")
        return self.index2word
```

File: gensim/models/keyedvectors.py [hunk 5]
```diff
@@ -336,6 +346,95 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
         result = [(self.index2word[sim], float(dists[sim])) for sim in best if sim not in all_words]
         return result[:topn]
 
+    def most_similar_among(self, positive=[], negative=[], topn=10, restrict_vocab=None, indexer=None,
```
Comment (shubhvachher) [inline gensim/models/keyedvectors.py]: For lack of imagination : <words_list> !

File: gensim/models/keyedvectors.py (hunk 1 above)
Comment (shubhvachher) [inline gensim/models/keyedvectors.py]: Ok, done. thanks for the catch.

File: gensim/models/keyedvectors.py (hunk 2 above)
Comment (shubhvachher) [inline gensim/models/keyedvectors.py]: Sure, done. Not storing `limited` anymore...
```python
    words_list_indices = [self.vocab[word].index for word in words_to_use]
            # limited = self.syn0norm[words_list_indices] #syn0norm is an ndarray; storing limited might add a huge memory overhead
        else:
            raise ValueError("words_list must be set/list of words. Please read doc string")

        dists = dot(self.syn0norm[words_list_indices], mean)
        result = []

```

File: gensim/models/word2vec.py (hunk 3 above)
Comment (shubhvachher) [inline gensim/models/word2vec.py]: hmm, I was confused at first myself but then I just did it to maintain coding style of word2vec.

Do you think it might be confusing for newcomers if we don't have the forwarding methods. It took me quite a while to figure KeyedVectors as having all after training functionality as word2vec!

Also, most of the word2vec functions don't have `docstring` and one has to read the `docstring` of w2v.wv to figure things out. That might throw people off. Would you like me to add docstrings for forwarding functions?

**Comment by tmylk** [comment]
Thanks for the feedback! Yes, having more docstrings is very useful. Explaining what functions do and forwarding to KeyedVectors for more

**Comment by shubhvachher** [comment]
My god. Unit testing is going to be comprehensive(CBOW, SG, cosmul?). I'll committed the changes until now. Adding unit tests soon.

**Comment by shubhvachher** [comment]
@tmylk Will do. Also, what do you think about @gojomo's last review suggestion? Should I add forwarding functions to the Word2Vec file for new functions in KeyedVectors or are we going to ask people to use the KeyedVectors class always?

**Comment by tmylk** [comment]
Please avoid adding new forwarding functions. KeyedVectors is the way to go.

**Comment by shubhvachher** [comment]
Those were my first unit tests! Please do review!

**Comment by shubhvachher** [comment]
I've only added docstrings in the first commit to fail tests. Have any idea what broke? Travis says : 
```
======================================================================

FAIL: testPipeline (gensim.test.test_sklearn_integration.TestSklearnLDAWrapper)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/RaRe-Technologies/gensim/gensim/test/test_sklearn_integration.py", line 108, in testPipeline

    self.assertGreater(score, 0.50)

AssertionError: 0.5 not greater than 0.5

----------------------------------------------------------------------
```

I haven't changed anything though but I found this within the test_word2vec.py file **(not the test that broke)** : 
```python

if not hasattr(TestWord2VecModel, 'assertLess'):
    # workaround for python 2.6
    def assertLess(self, a, b, msg=None):
        self.assertTrue(a < b, msg="%s is not less than %s" % (a, b))

    setattr(TestWord2VecModel, 'assertLess', assertLess)

```

Do you think something similar is needed?

**Comment by tmylk** [comment]
Thanks for reporting this. Please change the test in `testPipeline` to be   `self.assertGreater(score, 0.40)`. It is a new test.

**Comment by shubhvachher** [comment]
Ok, could you please also check #1233. If proper Ill add that fix here as well

**Comment by shubhvachher** [comment]
added fix for #1233

**Comment by shubhvachher** [comment]
This is done, when free could you please look this over @gojomo ? Thanks!

File: gensim/test/test_word2vec.py [hunk 6]
```diff
@@ -461,6 +462,141 @@ def test_cbow_neg(self):
                                   min_count=5, iter=10, workers=2, sample=0)
         self.model_sanity(model)
 
+    def test_most_similar_among_CBOW(self):
```
Comment (gojomo) [inline gensim/test/test_word2vec.py]: Because `most_similar_among()` functionality is merely on a set of vectors, it's not dependent on the mode-of-training – so there's no need for multiple models built different ways. Testing against one vector set, trained if necessary in the most simple/default way, would be sufficient. (In fact, it'd ultimately make even more sense to be inside a set of KeyedVectors tests, rather than Word2Vec tests, but since its sibling methods are tested here it's not necessary to make that big change yet.)

File: gensim/test/test_word2vec.py [hunk 7]
```diff
@@ -461,6 +462,141 @@ def test_cbow_neg(self):
...
+        """
+        Testing most_similar_among method for CBOW model
+        """
+
+        #Testing Error
+        model = word2vec.Word2Vec(sentences, size=2, sg=0, min_count=1, hs=1, negative=0)
+        self.assertRaises(ValueError, model.wv.most_similar_among, positive=['graph'])
+
+        res_voc = model.wv.index2word[:5] #Gives first 5 words in model vocab
+        #Testing Warnings
```
Comment (gojomo) [inline gensim/test/test_word2vec.py]: As noted previously, I think the `suppress_warnings` toggle adds unnecessary complication. Also, that logging may be more appropriate than warnings.

File: gensim/test/test_word2vec.py [hunk 8]
```diff
@@ -461,6 +462,141 @@ def test_cbow_neg(self):
...
+    def test_most_similar_among_CBOW(self):
+        """
+        Testing most_similar_among method for CBOW model
+        """
+
+        #Testing Error
+        model = word2vec.Word2Vec(sentences, size=2, sg=0, min_count=1, hs=1, negative=0)
+        self.assertRaises(ValueError, model.wv.most_similar_among, positive=['graph'])
+
+        res_voc = model.wv.index2word[:5] #Gives first 5 words in model vocab
```
Comment (gojomo) [inline gensim/test/test_word2vec.py]: Usual python style is two-spaces before an end-of-line `#`-comment, and a space after the `#` before text. (Applies many other places as well.)

**Comment by shubhvachher** [comment]
Thank you for the comprehensive review @gojomo ! I will work on this asap.

**Comment by tmylk** [comment]
@shubhvachher Please continue this pr instead of opening a new one in #1255.

**Comment by shubhvachher** [comment]
Ah, sorry. I am working on develop here; will remember not to do that in the future!

So, changes made : Moved drive by fixes into separate PRs and replied to comments above. rebased uptil now.

File: gensim/models/keyedvectors.py [hunk 9]
```diff
@@ -336,6 +346,106 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
         result = [(self.index2word[sim], float(dists[sim])) for sim in best if sim not in all_words]
         return result[:topn]
 
+    def most_similar_among(self, positive=[], negative=[], topn=10, words_list=None, indexer=None,
+                            suppress_warnings=False):
```
Comment (shubhvachher) [inline gensim/models/keyedvectors.py]: Actually, calculating difference in sets (which is needed to generate this warning) can be an unnecessary overhead for people using this code in a production environment. I have added a `logging.info` call to inform users of the same.
Sometimes it is not possible to remove all non-vocabulary words from the `words-list`, for example, if `words-list` is generated during runtime. Here using `suppress_warnings` would be ideal.

File: gensim/models/keyedvectors.py [hunk 10]
```diff
@@ -336,6 +346,106 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+            raise ValueError("words_list must be a set/list of words. Maybe you wanted the most_similar function.")
+        elif isinstance(words_list, list) or isinstance(words_list, set):
+            words_list = set(words_list)
+            vocabulary_words = set(self.index2word)
+
+            words_to_use = vocabulary_words.intersection(words_list)
+
+            if not suppress_warnings:
+                missing_words = words_list.difference(vocabulary_words)
+                if(not missing_words): # missing_words is empty
```
Comment (shubhvachher) [inline gensim/models/keyedvectors.py]: Changed for every occourance

File: gensim/models/word2vec.py [hunk 11]
```diff
@@ -355,6 +355,8 @@ class Word2Vec(utils.SaveLoad):
     """
     Class for training, using and evaluating neural networks described in https://code.google.com/p/word2vec/
 
+    If you're finished training a model (=no more updates, only querying), then switch to the :mod:`gensim.models.KeyedVectors` instance in wv
```
Comment (shubhvachher) [inline gensim/models/word2vec.py]: Tried following pep8, please do check now

File: gensim/models/word2vec.py [hunk 12]
```diff
@@ -1181,33 +1187,83 @@ def intersect_word2vec_format(self, fname, lockf=0.0, binary=False, encoding='ut
         logger.info("merged %d vectors into %s matrix from %s" % (overlap_count, self.wv.syn0.shape, fname))
 
     def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, indexer=None):
+        """
+        Please refer to the documentation for `gensim.models.KeyedVectors.most_similar`
+        In the future please try and use the `gensim.models.KeyedVectors` instance in wv
```
Comment (shubhvachher) [inline gensim/models/word2vec.py]: Changed this and split into a separate PR!

File: gensim/test/test_sklearn_integration.py [hunk 13]
```diff
@@ -105,7 +105,7 @@ def testPipeline(self):
         text_lda = Pipeline((('features', model,), ('classifier', clf)))
         text_lda.fit(corpus, data.target)
         score = text_lda.score(corpus, data.target)
-        self.assertGreater(score, 0.50)
+        self.assertGreater(score, 0.40)
```
Comment (shubhvachher) [inline gensim/test/test_sklearn_integration.py]: Split into a separate PR! Thank you so much. Taught me a lot about git and got me my first commit :)

File: gensim/test/test_word2vec.py (hunk 6 above)
Comment (shubhvachher) [inline gensim/test/test_word2vec.py]: Sure, Removed extra tests; using CBOW method to check functionality.

That second part might be important given @tmylk's comment
> KeyedVectors is the way to go.

What do you think @tmylk ?

File: gensim/test/test_word2vec.py (hunk 7 above)
Comment (shubhvachher) [inline gensim/test/test_word2vec.py]: logging is used. Suppress warnings becomes useful in case `words-list` is generated on the go and the user does not want to know about missing words.

File: gensim/models/word2vec.py (hunk 3 above)
Comment (shubhvachher) [inline gensim/models/word2vec.py]: I have added only this method to class `Word2Vec` in order to solve the original "people might not find index2word" problem.

File: gensim/models/keyedvectors.py [hunk 14]
```diff
@@ -338,6 +345,129 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+        else:  # This is triggered for empty words_list parameter
+            raise ValueError("words_list must be set/list of words. " \
+                                "Maybe you wanted the most_similar function. " \
+                                "Please read doc string")
+
+        if type(topn) is not int:
+            if topn is False:
+                pass
+            else:
+                if suppress_warnings is False:
```
Comment (tmylk) [inline gensim/models/keyedvectors.py]: This must be an exception, not a warning. Incorrect input can't be surpressed.

File: gensim/models/keyedvectors.py [hunk 15]
```diff
@@ -338,6 +345,129 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+            missing_words = words_list.difference(vocabulary_words)
+            if not missing_words:  # missing_words is empty
+                pass
+            else:
+                logger.warning("The following words are not in " \
+                                "trained vocabulary : %s", str(missing_words))
+                logger.info("This warning is expensive to calculate, " \
+                                "especially for large words_list. " \
+                                "If you would rather not remove the missing_words " \
+                                "from words_list please set the " \
```
Comment (tmylk) [inline gensim/models/keyedvectors.py]: Better message is "Please intersect with vocabulary words_to_use = vocabulary_words.intersection(words_list) prior to calling the most_similar_among". 
Please remove the `suppress_warnings` flag.

File: gensim/models/keyedvectors.py [hunk 16]
```diff
@@ -338,6 +345,129 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+                                "suppress_warnings flag.")
+
+        words_list_indices = [self.vocab[word].index for word in words_to_use]
+        # limited = self.syn0norm[words_list_indices]
+        # Storing 'limited' might add a huge memory overhead so we avoid doing that
```
Comment (tmylk) [inline gensim/models/keyedvectors.py]: Please memory profile this code to provide foundation for this statement.
Please remove commented out code.

File: gensim/models/keyedvectors.py [hunk 17]
```diff
@@ -338,6 +345,129 @@ def most_similar(self, positive=[], negative=[], topn=10, restrict_vocab=None, i
...
+
+        Example::
+
+          >>> trained_model.most_similar_among(positive=['man'], topn=1,
+                                                words_list=['woman','random_word'])
+          [('woman', 0.75882536)]
+
+        """
+
+        if isinstance(words_list, int):
```
Comment (tmylk) [inline gensim/models/keyedvectors.py]: Please use a single check and a single  `raise ValueError`. 
The `most_similar` function doesn't take a list of `int`s so it should not be mentioned here

**Comment by tmylk** [comment]
Ping @shubhvachher . It is a useful functionality, would be good to finish this PR

**Comment by shubhvachher** [comment]
Yup, sorry! Finishing up exams here. On this asap.

**Comment by menshikh-iv** [comment]
Ping @shubhvachher, what status of this PR? Will you finish it soon?

**Comment by shubhvachher** [comment]
Yes @menshikh-iv; my bad. Been travelling and working. I do have access to my laptop though. I'd get on this coming weekend if thats ok!

**Comment by menshikh-iv** [comment]
Ok @shubhvachher, we will wait.

**Comment by menshikh-iv** [comment]
ping @shubhvachher

**Comment by menshikh-iv** [comment]
I close current PR because this looks abandoned.

**Comment by shubhvachher** [comment]
This is surely my fault for lack of time from my summer work and travels; Will open this again as soon as I get the chance and if uncleared. Apologies,
Shubh

**Comment by piskvorky** [comment]
@shubhvachher please do. The (semi-automated) closing of stale tickets is a maintenance thing from our side, not an indication we're not interested in the PR!
